storescu/src/main.rs

```diff
@@ -313,190 +351,167 @@ fn run() -> Result<(), Error> {
+                saml_assertion,
+                jwt,
+                pc,
+            )
+            .await?;
+            loop {
+                let file = {
+                    let mut files = d_files.lock().await;
+                    files.pop()
+                };
+                let mut file = match file {
+                    Some(file) => file,
+                    None => break,
+                };
+                let r: Result<_, Error> = check_presentation_contexts(
+                    &file,
+                    scu.presentation_contexts(),
+                    never_transcode,
+                );
+                match r {
+                    Ok((pc, ts)) => {
+                        if verbose {
+                            debug!(
+                                "{}: Selected presentation context: {:?}",
+                                file.file.display(),
+                                pc
                             );
-                            if fail_first {
-                                let _ = scu.abort();
-                                std::process::exit(-2);
-                            }
                         }
-                        _ => {
-                            error!(
-                                "Failed to store instance `{}` (status code {:04X}H)",
-                                storage_sop_instance_uid, status
-                            );
-                            if fail_first {
-                                let _ = scu.abort();
-                                std::process::exit(-2);
-                            }
+                        file.pc_selected = Some(pc);
+                        file.ts_selected = Some(ts);
+                    }
+                    Err(e) => {
+                        error!("{}", Report::from_error(e));
+                        if fail_first {
+                            let _ = scu.abort().await;
+                            std::process::exit(-2);
                         }
                     }
                 }
-
-                pdu @ Pdu::Unknown { .. }
-                | pdu @ Pdu::AssociationRQ { .. }
-                | pdu @ Pdu::AssociationAC { .. }
-                | pdu @ Pdu::AssociationRJ { .. }
-                | pdu @ Pdu::ReleaseRQ
-                | pdu @ Pdu::ReleaseRP
-                | pdu @ Pdu::AbortRQ { .. } => {
-                    error!("Unexpected SCP response: {:?}", pdu);
-                    let _ = scu.abort();
-                    std::process::exit(-2);
-                }
+                scu = send_file(scu, file, message_id, pbx.as_ref(), verbose, fail_first).await?;
+            }
+            let _ = scu.release().await;
+            Ok::<(), Error>(())
+        });
+    }
+    while let Some(result) = tasks.join_next().await {
+        if let Err(e) = result {
+            error!("{}", Report::from_error(e));
+            if fail_first {
+                std::process::exit(-2);
             }
```

Comment:
This code is testing the error on task join, but not the application error returned by the task. This worked better on my machine.
```suggestion
        match result {
            Err(e) => {
                error!("{}", Report::from_error(e));
                if fail_first {
                    std::process::exit(-2);
                }
            }
            Ok(Err(e)) => {
                error!("{}", Report::from_error(e));
                if fail_first {
                    std::process::exit(-2);
                }
            }
            Ok(Ok(_)) => {}
        }
```